group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java

```diff
@@ -498,29 +497,17 @@ public CompletableFuture<ListGroupsResponseData> listGroups(
             );
         }
 
-        final Set<TopicPartition> existingPartitionSet = runtime.partitions();
-
-        if (existingPartitionSet.isEmpty()) {
-            return CompletableFuture.completedFuture(new ListGroupsResponseData());
-        }
-
-        final List<CompletableFuture<List<ListGroupsResponseData.ListedGroup>>> futures =
-            new ArrayList<>();
-
-        for (TopicPartition tp : existingPartitionSet) {
-            futures.add(runtime.scheduleReadOperation(
-                "list-groups",
-                tp,
-                (coordinator, lastCommittedOffset) -> coordinator.listGroups(request.statesFilter(), request.typesFilter(), lastCommittedOffset)
-            ).exceptionally(exception -> {
-                exception = Errors.maybeUnwrapException(exception);
-                if (exception instanceof NotCoordinatorException) {
-                    return Collections.emptyList();
-                } else {
-                    throw new CompletionException(exception);
-                }
-            }));
-        }
+        final List<CompletableFuture<List<ListGroupsResponseData.ListedGroup>>> futures = runtime.scheduleReadAllOperation(
+            "list-groups",
+            (coordinator, lastCommittedOffset) -> coordinator.listGroups(request.statesFilter(), request.typesFilter(), lastCommittedOffset)
+        ).stream().map(future -> future.exceptionally(exception -> {
```

Comment:
how about adding the exception function to `scheduleReadAllOperation`? It brings two benefits:

1. simplify the code from callers 
2. avoid creating list repeatedly

Reply:
I actually considered this but eventually rejected it because those new methods could also be used with a different handling than `exceptionally`. The separation of concerns would not be too good in my opinion. I was also considering adding an helper like `mapExceptionally` to add an exception handler on a list of futures. Do you think that it could also help?

Reply:
> adding an helper like mapExceptionally to add an exception handler on a list of futures

Do you mean to enhance `FutureUtils#combineFutures` to take one more parameter? Current PR is good to me but I'd like to see more design  (`mapExceptionally`) before merging :)

Reply:
I just pushed a commit with the `mapExceptionally` construct that I was thinking about. Please let me know what you think.